single_kernel_mongo/state/abstract_state.py

```diff
@@ -0,0 +1,57 @@
+# Copyright 2024 Canonical Ltd.
+# See LICENSE file for licensing details.
+"""The charm state for mongo charms (databags + model information)."""
+
+from typing import Generic, TypeVar
+
+from ops.model import Application, Relation, Unit
+
+from single_kernel_mongo.config.literals import Substrates
+from single_kernel_mongo.lib.charms.data_platform_libs.v0.data_interfaces import (  # type: ignore
+    Data,
+)
+
+PData = TypeVar("PData", bound=Data, covariant=True)
+
+
+class AbstractRelationState(Generic[PData]):
+    """Relation state object."""
+
+    def __init__(
+        self,
+        relation: Relation | None,
+        data_interface: PData,
+        component: Unit | Application | None,
+        substrate: Substrates | None = None,
+    ):
+        self.relation = relation
+        self.data_interface = data_interface
+        self.component = component
+        self.substrate = substrate
+        self.relation_data = self.data_interface.as_dict(self.relation.id) if self.relation else {}
+
+    def __bool__(self) -> bool:
+        """Boolean evaluation based on the existence of self.relation."""
+        try:
+            return bool(self.relation)
+        except AttributeError:
+            return False
+
+    def update(self, items: dict[str, str]) -> None:
```

Comment:
@Gu1nness curious what your thoughts are here

Reply:
It's a good question actually.
I love in `update` that it's a single entrypoint to do any update in the databag, and it does not need to know what is in the databag to update it.
But you're right that we could be more CRUD like.
I have no strong opinion on that TBH.

Reply:
No, it will not delete the other fields: it deletes the fields where you pass `None` as a value (like we do in secrets btw, if you pass None, it will delete the secret field, if the secret field was not present we add it, and if it was present, we update the value)

Reply:
Is it ok to keep it in the backlog for later or do you want me to change this right now?
The idea behind this update is that: you don't need to know if the field exists to set it and anyway the value would be overwritten by what you write.
I can for a first step have a `set` and `delete`, would that make it ok for you?

Reply:
We can keep it for now! thanks for making that clear! It wasn't so obvious when I first read it but now it is.

Instead of a docstring `        """Writes to relation_data."""`. maybe it could be 
```
"""
Adds/Updates/Deletes fields.

It deletes the fields where you pass None as a value, if the value doesn't exists then we add it, and if it was present, we update the value.
"""
```